### PR TITLE
feat: allow users to search for commands

### PR DESCRIPTION
**Pull request by Pop0097**
This feature allows users the ability to search for commands using command line arguments. 

- `-a` for alias
- `-t` for tag
- `-c` for command

We search and filter in the following order
1. Tag
2. Command
3. Alias

Once the user selects a command, it is added to their clipboard and the last used property of the command is updated in the database. 

https://github.com/danyal002/cmd-stack/assets/52176616/90475bd1-cd32-4128-a732-92aaa024e6f2

### CODE REVIEW

File: cli/cmd/delete.go
Comment (danyal002) [inline cli/cmd/delete.go:32]: We should decide on a variable naming convention and stick to it. If we're going to use camel case, then we should change it everywhere in future PRs.

File: cli/cmd/search.go
Comment (danyal002) [inline cli/cmd/search.go]: Hmm, I don't like this. This introduces more edge cases. We should either have a `nil` returned or an actual value. What do you think?

(This is coming from me programming in Rust and getting annihilated for doing the above in my PRs)

File: cli/cmd/search.go
Comment (danyal002) [inline cli/cmd/search.go]: why the `cur_` prefix?

File: cli/cmd/search.go
Comment (danyal002) [inline cli/cmd/search.go]: You don't have to do this in this PR, but this should be a results object, something like `SearchResult`.

File: cli/cmd/search.go
Comment (danyal002) [inline cli/cmd/search.go]: Imo, the "command" or `-c` should be the default argument for the search command e.g. `cmdstack search "some command" -t ... -n ...` and so on.

File: cli/cmd/search.go
Comment (danyal002) [inline cli/cmd/search.go]: I think there's a better way to do this than what's below, but I can't remember off the top of my head. I remember reading somewhere that you can autogenerate this part using Cobra.

File: cli/cmd/search.go
Comment (danyal002) [inline cli/cmd/search.go]: Confusing name. A better name would be `FilterCommandsByCommandText` or `FilterCommandsByCommandContent`.

File: cli/cmd/search.go
Comment (danyal002) [inline cli/cmd/search.go]: I'd remove the "Select Command" prefix. The little triangle on the left side of every command in the list is enough to know what the user is doing. It is also less cognitive load.

File: cli/dal/command.go
Comment (danyal002) [inline cli/dal/command.go:54]: Hmm, whilst this is fine for now, this will quickly become a maintenance issue because we'll be reusing it everywhere. I would suggest searching for a library that can take care of this for us. I'm sure one exists.

In Rust, we used something called `Validator` which would use attributes on each struct property to validate a struct every time it was created.

File: cli/dal/dal.go
Comment (danyal002) [inline cli/dal/dal.go]: nit: `Failed to execute query:`.

File: cli/dal/dal.go
Comment (danyal002) [inline cli/dal/dal.go:1]: So, one big overarching comment I have is that I don't agree with your approach for search and filtering. A database is much better equipped to handle that for us. We're reinventing the wheel here. The way I would do this is by creating a `SearchFilters` object. You would then pass that to the `dal` `search` method and form your query using the filters. This is where `Squirrel` can help. You would validated the filters object there and return an error if the filters are invalid. Execute the query and create a `SearchResult` object to return.

**Review by danyal002 — CHANGES_REQUESTED** [review]

File: cli/cmd/search.go
Comment (Pop0097) [reply cli/cmd/search.go]: I agree, however with the way that the UI is generated by the library you need a prompt otherwise a `:` just appears.

File: cli/cmd/search.go
Comment (Pop0097) [reply cli/cmd/search.go]: Can you? I had to use an external library for this

File: cli/dal/command.go
Comment (Pop0097) [reply cli/dal/command.go:54]: Will do in future PR

File: cli/cmd/search.go
Comment (Pop0097) [reply cli/cmd/search.go]: Can address in future PR

File: cli/cmd/search.go
Comment (Pop0097) [reply cli/cmd/search.go]: Will remove

File: cli/cmd/search.go
Comment (Pop0097) [reply cli/cmd/search.go]: This is a good catch, thanks for the suggestion!

File: cli/cmd/delete.go
Comment (Pop0097) [reply cli/cmd/delete.go:32]: camel case

File: cli/cmd/list.go
Comment (danyal002) [inline cli/cmd/list.go]: Nit: could this be better named? I'm not sure what the difference between `resultLimit` and `printLimit` is.

File: cli/cmd/search.go
Comment (danyal002) [inline cli/cmd/search.go:28]: We should talk about setting up a config file for all these settings.

File: cli/cmd/search.go
Comment (danyal002) [inline cli/cmd/search.go]: This would be better as a method I think, unless you plan on using it without a `Command`.

**Review by danyal002 — COMMENTED** [review]
Could you move the `example_commands.txt` file to a separate `test_data` directory please?

File: cli/cmd/search.go
Comment (Pop0097) [reply cli/cmd/search.go]: Do you mean a method of the DAL?

Comment (danyal002) [reply cli/cmd/search.go]: the `Command` object as it's always called on a `Command`.

**Review by danyal002 — APPROVED** [review]
Approving with one comment

**Pull request merged**
